### ISSUE TITLE
Turn indexing with a scalar tensor into an copy into a view and avoid a D2H synchronization.

### ISSUE THREAD

**Issue by Chillee** [pytorch/pytorch#105641]
### 🚀 The feature, motivation and pitch

Today, this triggers a cuda synchronization:
```
import torch

torch.set_default_device('cuda')

def f(x, y):
    return x[y]

inps = (torch.randn(5), torch.tensor(0))
torch.cuda.set_sync_debug_mode(2)
f(*inps)
```

The reason why is that when the tensor is a 0-dim value, instead of launching a gather kernel, we move the tensor to the hsot and do a slice instead (https://github.com/pytorch/pytorch/pull/105518/files#diff-2574bfb0ffa78d685fb7bd2ebc0c64b1a5f87dd55ec74ae67b41b31adc566020L466).

We could just fix this, but unfortunately, this does change the semantics. In particular, now, this operation would create a copy instead of a view, which could cause issues for downstream in-place operations.

I think these are bad semantics, for 2 reasons:

1. Cuda synchronizations are very bad in general. They're slow, prevent the use of many different features (streams, cudagraphs, don't play well with collectives, etc.), and should strongly be avoided. This, however, is a very implicit coercion we're doing. It's not obvious at all that if the tensor is 3-dim/2-dim/1-dim it doesn't do a sync, but if the tensor is 0-dim it does do a sync.

2. Moreover, it's *not* consistent!!

Why should `x[torch.tensor(0)]` return a view but `x[torch.tensor([0])` return a copy? Why should the first one do a synchronization and the second one doesn't?

To drive this point home further, we also diverge from Numpy semantics here.

```
import numpy as np
x = np.ones(5)
y = np.array(1)
z = x[y]
z += 1
print(x)
>>> array([1., 1., 1., 1., 1.])
```

PS: I've also done a brief survey of use cases with this pattern I could find (https://github.com/search?q=%2F%28.%2B%29%5C%5Btorch.tensor%5C%28%2F&type=code), and all of them seem like they would *benefit* from these semantics being change, particularly if `torch.set_default_device('cuda')` is set to true. I will go more into detail in a follow-up comment.

**Comment by lezcano** [comment]
See https://github.com/pytorch/pytorch/pull/105518#discussion_r1269173619 for a workaround

**Comment by zou3519** [comment]
+1 for the inconsistency with numpy. [numpy's docs say advanced indexing always returns a copy](https://numpy.org/doc/stable/user/basics.indexing.html#advanced-indexing) (instead of a view!), but we are returning a view in the scalar tensor case.

**Comment by ezyang** [comment]
OK, maybe we can chalk this up as a bug fix. When you changed the behavior in your PR, did a lot of tests start failing?

**Comment by Chillee** [comment]
@ezyang haven't fixed all the tests yet, but no. In fact, most of the tests that started failing were previously failing tests that were now passing (unexpected success), since they didn't successfully trace previously.

**Comment by lezcano** [comment]
Actually, here we're modelling NumPy as well:
```python
>>> x = np.zeros(3, dtype=np.float32)
>>> u = np.zeros((), dtype=np.int64)
>>> x[u] += 1
>>> x
array([1., 0., 0.], dtype=float32)
```

Those docs do not seem entirely correct.

**Comment by Chillee** [comment]
@lezcano Numpy scalars are weird 🤔 But even when it returns a numpy array, it's also not a view.

To be clear, `x[u] += 1` dispatches to setitem, which is a totally different codepath.

**Comment by zou3519** [comment]
To summarize the discussion a bit:
- YOLO land is fine assuming we do not not hit the BC-breaking case (in-place into advanced indexing with scalar tensors) in PyTorch tests or in internal tests (that would imply that this case matters more than we think)
- We should change the behavior for CPU as well to make everything consistent with numpy

**Comment by Chillee** [comment]
> We should change the behavior for CPU as well to make everything consistent with numpy

I'm actually a bit worried about this, since if the indexing tensor is on CPU this change is no longer necessarily perf positive.

**Comment by colesbury** [comment]
I've chatted with @Chillee and @zou3519 about this, but I think it would be worthwhile to summarize my thoughts here.

The proposed change poses serious backwards compatibility issues. I do not think we should go forward with it, especially since we have alternatives that avoid D2H synchronization without breaking backwards compatibility.

### Backwards compatibility

People frequently write code that looks like:

```
tensor[idx][idx2] = value
```

or

```
tensor[idx].zero_()
```

If `idx` is a zero-dim Tensor, then this code will no longer work as expected. @Chillee found https://github.com/pytorch/pytorch/issues/6217 where we ran into this when we first implemented advanced indexing, and had to block the 0.4 release to change the behavior.

Here are a few examples of code in the wild that will break:
* https://github.com/open-intelligence/OpenI-Octopus/blob/6c7f440e5fb8e564feaa5fb72f86b23fa68b294b/model-compression/prune/pruning.py#L117 (found by @Chillee)
 * https://github.com/ICT-ANS/StarLight/blob/c23e77f144f78ae93d630bf5bafd09fb552bc0f5/algorithms/nas/DDSAS/models/model_search.py#L183

### NumPy

PyTorch zero-dim tensors are a substitute for both NumPy zero-dim arrays and NumPy scalars. In practice, **it is typically more important to behave like NumPy scalars** because zero-dim arrays are rare in NumPy, whereas scalars frequently arise from indexing a 1d array. For example, the below code currently behaves equivalently in NumPy and PyTorch, but would behave differently if we implemented the behavior described in this issue

```python
indices = np.array([0, 10, 20, 30])
array = np.random.randn(100, 100)
slice = array[indices[0]]
slice[:] = 0  # writes back to array
```

```python
indices = torch.tensor([0, 10, 20, 30])
array = torch.randn(100, 100)
slice = array[indices[0]]
slice[:] = 0   # writes back to array
```

### Alternatives

1. Expose `Tensor::index` to Python. This implements advanced indexing and always returns a copy.
2. Convert zero-dim indexes to 1-element 1-d indexes.

**Comment by ezyang** [comment]
@kurtamohler I wonder if CoW could help us here

**Comment by kurtamohler** [comment]
Yeah, if the plan is to change advanced indexing to always return a clone instead of a view, we could return a COW (lazy clone) tensor to make the operation more performant than if we returned an immediate clone.

Of course, this would still be BC-breaking, since lazy clone is semantically the same as a clone. If we decide that's alright, we could use the same deprecation cycle strategy discussed in https://github.com/pytorch/pytorch/issues/109833#issuecomment-2073651428, to detect and warn on any cases where numerical results will differ in the future
